Streamline WCS fitting

ceb8 opened the pull request:
This pertains to TESS cutouts.
Currently when fitting a linear WCS to a cutout we use every single pixel in the fit. This gets inefficient for large cutouts and does not make the fit appreciable better.
This PR limits the number of pixels used to a maximum of 100, spread evenly throughout the cutout.

scfleming commented on astrocut/cube_cut.py:
```diff
@@ -272,17 +272,26 @@ def _fit_cutout_wcs(self, cutout_wcs, cutout_shape):
+
+        y, x = cutout_shape
+        i = 1
+        while (x/i)*(y/i) > 100:
```
Suggest adding an in-line comment on what you are doing here.  Took me a while to figure it out (if I got it right), I think this line is determining how large the spacing should be to ensure 100 points are sampled across the (user-specified) cutout along the x and y dimension?  Does this make any implict assumption about it being a square vs. an extreme rectangle?  E.g., does this work for a requested cutout that is 1 pixel in x-dimension and 50 pixels in the y-dimension?

scfleming replied:
The fact that the while loops over both the x- and y-dimensions with a single index counter ("i") is why I wonder if it works as expected for rectangles.  If the dimension is not a square, does this technique require looping over two different index counters to account for the spacing being different between the two dimensions?  I'd have to try it out myself with a stand-alone script for sure, but wanted to document my thought on this here.

ceb8 replied:
I believe this is working as intended and I will certainly add a comment. The goal is to create an evenly spaced mesh of no more than 100 pixel coordinates to use in the fit. In practice I do this by taking the ith value in each row/column.

This is what  happens for a 5x7 cutout with i = 2:
```
xoxoxox
ooooooo
xoxoxox
ooooooo
xoxoxox
```

Where x denotes the indexes that will be used in the fit.

Does that make sense?

scfleming replied:
Yes it does, in fact this exact thing, including the tic-tac-toe board, would be a great thing to add as an in-line comment block.

ceb8 replied:
Will do!

scfleming commented on astrocut/cube_cut.py:
```diff
@@ -272,17 +272,26 @@ def _fit_cutout_wcs(self, cutout_wcs, cutout_shape):
         self.cutout_wcs = linear_wcs
 
-        # Checking the fit
+        # Checking the fit (we want to use all of the pixels for this)
```
What is being checked here?  Is "dists" used anywhere else?

ceb8 replied:
Nope, dists is the angular distance between the true skycoords as calculated with the original wcs and the approximate skycoords as calculated with the linear wcs.

ceb8 replied:
This is meant to be a measure of how good the approximate wcs is. It's very good for small cutouts and gets progressively worse for larger cutouts.

ceb8 commented:
Here are plots for four different parts of the sky: near the pole, near the equator, over the 0->360 crossover, and a random part of the sky I've been coding off of. I tested a variety of shapes/sizes from 5x5 up to 200x200 (the maximum cutout size the API allows). I also testes cutouts in the corner of the FFI, near the center, and halfway between.

At the 0->360 crossover:
![354_66_100_plot](https://user-images.githubusercontent.com/9406508/80996953-a8bf0a00-8e0e-11ea-93a3-088d9dde777a.png)

Near the equator:
![125_7_100_plot](https://user-images.githubusercontent.com/9406508/80997036-cab88c80-8e0e-11ea-8f0a-1374e42d84dd.png)

Near the pole:
![104_-74_100_plot](https://user-images.githubusercontent.com/9406508/80997103-dc9a2f80-8e0e-11ea-8f8f-6c5cecf0fc45.png)

Random sky position:
![259_36_100_plot](https://user-images.githubusercontent.com/9406508/80997142-e7ed5b00-8e0e-11ea-9ef8-6b3b86e60374.png)

Note how mostly the error tracks the size of the cutout, increasing slightly faster when using a maximum of 100 points, but there are exceptions, interestingly in both directions where either the new fit is better or the old one is. 

Also, while making these plots I noticed some issues with the crpix values sometimes being really off. Interestingly this did not correspond to higher error fits, and also was a problem for both the new and old fits. I switched to using "center" for the center coordinate rather than giving a specific SkyCoord and that seems to have solved the problem. I originally did not do that because it failed at the 0->360 degree crossover, a bug which appears to have been fixed.  The fact that giving a SkyCoord as the center value sometimes gives anomalous results is a bug in the fitting code that should be resolved in astropy, however as using "center" does work, that does not need to be fixed for this PR. (I plan to explore this further and put in an issue/PR with astropy).

ceb8 commented:
OK I did more digging and made some more changes and improved the fit in a number of ways:
- Created the grid of pixels centered on the center pixel and made sure the the edge pixels are included
  This makes the output naxis values correct and, makes the input grid symmetrical and even which seems to make the fit go faster
- Added bounds in the wcs fitting code so that the CRPIX value must be on the cutout
  This takes care of the oddly bad fits I was sometimes getting
- Added 1 to the pixel shape in the wcs fitting code
  This code was 1 off because it was subracting pixel indices rather than lengths

The wcs fitting updates will go into a PR for astropy.

Here are the updated plots for the different parts of the sky, where we can see that the fits are actually improved through these changes.

0->360 crossover
![354_66_100_plot](https://user-images.githubusercontent.com/9406508/81104948-4125bf00-8ee1-11ea-8168-f0ca6a1c5f52.png)

Near the equator:
![125_7_100_plot](https://user-images.githubusercontent.com/9406508/81105021-5a2e7000-8ee1-11ea-8175-eaa0714ea489.png)

Near the pole:
![104_-74_100_plot](https://user-images.githubusercontent.com/9406508/81105039-60bce780-8ee1-11ea-8779-d101bd78e0c7.png)

Random sky position:
![259_36_100_plot](https://user-images.githubusercontent.com/9406508/81105071-67e3f580-8ee1-11ea-8362-3657b5278e52.png)
